Implement CVS fetcher

eschnett opened the pull request.

eschnett committed:
Implement CVS fetcher

eschnett committed:
Clean up code

eschnett commented:
The CVS fetcher isn't covered by self-tests. Where/how should I add them? Should I?

adamjstewart commented:
I think tests would be good, see `lib/spack/spack/test/(git|hg|svn)_fetch.py` for examples. Note that these tests are skipped if the dependencies aren't installed, so you'll need to install `cvs` in `.github/workflows/unit_tests.yaml` to actually get coverage.

Docs would also be good. They can be added here: https://spack.readthedocs.io/en/latest/packaging_guide.html#fetching-from-code-repositories

eschnett committed:
Implement CVS self tests

eschnett committed:
Beautify source code

eschnett commented:
The error is:
```
2021-04-23T20:57:43.3020151Z   >> 1226    Bus error (core dumped)
2021-04-23T20:57:43.3020728Z      1227    make: *** [Makefile:604: sharedmods] Error 135
2021-04-23T20:57:43.3021347Z      1228    make: *** Waiting for unfinished jobs....
```
while installing Python. That's unrelated to my pull request.

eschnett committed:
Add CVS fetching documentation

eschnett committed:
Create mock CVS repository

eschnett committed:
Beautify code

eschnett committed:
Create another mock CVS repo

eschnett committed:
Correct errors

eschnett committed:
Correct syntax error

eschnett committed:
Correct syntax error

eschnett committed:
Correct syntax error

eschnett committed:
Track down syntax error

eschnett committed:
Track down syntax error

eschnett committed:
Test CVS repositories

eschnett committed:
Beautify source code

eschnett committed:
Add 'date' keys

eschnett commented:
The last remaining error is from codecov, complaining that part of the code is not covered by tests. That's not true – these lines are definitively executed when running the unit tests. (I recall having to correct errors in some of these lines.)

eschnett commented:
This is ready to be reviewed, btw. Docs and tests are there as well.

eschnett commented:
Ping?

eschnett commented:
Ping?

adamjstewart commented on .github/workflows/unit_tests.yaml:
```diff
@@ -288,7 +288,7 @@ jobs:
       shell: runuser -u spack-test -- bash {0}
```
I think somewhere in this file you'll also want to `apt` or `dnf` install CVS, otherwise the tests will get skipped and there won't be any coverage.

adamjstewart commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1173,6 +1174,59 @@ you can check out a branch or tag by changing the URL. If you want to
+^^^
+
+CVS is an old centralized version control system similar to
+Subversion.
```
I would say something like "Subversion is a fork of an older centralized version control system called CVS."

eschnett replied:
Done.

adamjstewart commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1173,6 +1174,59 @@ you can check out a branch or tag by changing the URL. If you want to
+destination directory will be the standard stage source path.
+
+Fetching the head
+  Simply add an ``cvs`` parameter to the package:
```
```suggestion
  Simply add a ``cvs`` parameter to the package:
```

eschnett replied:
Done.

adamjstewart commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1173,6 +1174,59 @@ you can check out a branch or tag by changing the URL. If you want to
+
+         version('1.1.2.4')
+
+  CVS repositories locations are described using an older syntax that
```
```suggestion
  CVS repository locations are described using an older syntax that
```

eschnett replied:
Done.

adamjstewart commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -940,6 +940,126 @@ def __str__(self):
 
+@fetcher
+class CvsFetchStrategy(VCSFetchStrategy):
+
```
Delete this empty line

adamjstewart replied:
I think you deleted the wrong line, causing the style tests to fail.

eschnett replied:
Corrected.

adamjstewart commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -940,6 +940,126 @@ def __str__(self):
+
+           version('name',
+                   cvs=':pserver:anonymous@www.example.com:/cvsroot%module=modulename',
+                   branch='branchname', date='date')
```
Does CVS have branches? You didn't mention this feature in the docs

eschnett replied:
It does have branches. I updated the documentation.

adamjstewart commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -940,6 +940,126 @@ def __str__(self):
+    @property
+    def expect(self):
+        if not self._expect:
+            self._expect = which('expect', required=True)
```
Is `expect` the only way to run these commands? It seems to be pretty common on Linux/macOS so it might be fine, but if there were a way to avoid another external dependency that would be great. Speaking of which, you should add `cvs` (and possibly `expect`) to the Prerequisites listed in the Getting Started docs.

eschnett replied:
`expect` is necessary to enter a dummy password when connecting to a remote server. There doesn't seem to be a way to specify the password on the command line or on stdin.

adamjstewart replied:
Well, that's annoying, but if anyone needs to install CVS anyway they can always install expect if they need to.

adamjstewart commented on lib/spack/spack/test/cvs_fetch.py:
```diff
@@ -0,0 +1,92 @@
+
+
+pytestmark = pytest.mark.skipif(
+    not which('cvs'),
```
You should also check to make sure `expect` is installed if we need that for the tests

eschnett replied:
The tests use a local repository, and thus don't need `expect`.

adamjstewart reviewed (changes requested):
I'm guessing we aren't getting coverage because CVS isn't installed on the OS the GitHub Actions are running on.

eschnett committed:
Install CVS for self-tests

eschnett committed:
Update CVS documentation

eschnett committed:
Beautify source code

eschnett committed:
Beautify code

eschnett committed:
Update documentation

eschnett committed:
Add missing empty line

adamjstewart commented:
```
=================================== FAILURES ===================================
_______________________________ test_fetch[date] _______________________________

type_of_test = 'date'
mock_cvs_repository = <spack.util.pattern.Bunch object at 0x7ff91df251d0>
config = <spack.config.Configuration object at 0x7ff937b526d8>
mutable_mock_repo = <spack.repo.RepoPath object at 0x7ff91e6cca20>

    @pytest.mark.parametrize("type_of_test", ['default', 'date'])
    def test_fetch(
            type_of_test,
            mock_cvs_repository,
            config,
            mutable_mock_repo
    ):
        """Tries to:
    
        1. Fetch the repo using a fetch strategy constructed with
           supplied args (they depend on type_of_test).
        2. Check if the test_file is in the checked out repository.
        3. Assert that the repository is at the date supplied.
        4. Add and remove some files, then reset the repo, and
           ensure it's all there again.
        """
        # Retrieve the right test parameters
>       t = mock_cvs_repository.checks[type_of_test]
E       KeyError: 'date'

lib/spack/spack/test/cvs_fetch.py:43: KeyError
!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
== 1 failed, 642 passed, 10 skipped, 21 xfailed, 1 xpassed in 541.68 seconds ===
```

eschnett committed:
Make CVS tests succeed

eschnett committed:
Disable testing branches and dates for CVS

Branches, revisions, and dates are not unique; they are managed separately per file.

adamjstewart commented:
```
lib/spack/spack/test/cvs_fetch.py:46: [F841] local variable 'h' is assigned to but never used
```

eschnett committed:
Simplify code

adamjstewart reviewed (dismissed):
All of my requested changes have been integrated, so this looks good to me. There's still a lot of lines that aren't covered by tests, so that's the only remaining thing to improve. I'll let someone else review before merging.

eschnett commented:
I checked many of the lines that are not tested, and these seem to refer to checking out CVS repositories by branch or by date. These are difficult to test since CVS manages files independently, i.e. there is no unique notion of "the branch" or "the revision" of a checkout.

adamjstewart commented:
Do we need support for branches and dates for your application? If not we could just remove those features.

eschnett commented:
No, I don't need branches. Should I remove support for branches?

adamjstewart commented:
If it's not possible to test it I think it would be better to remove that feature. If you can find a way to test it, we can keep it.

eschnett committed:
Add tests for CVS branches and dates

eschnett committed:
Store date for each revision instead of using file system date

eschnett committed:
Don't import glob

eschnett committed:
Beautify code

eschnett committed:
Beautify code

eschnett commented:
I added some tests for checking out branches and dates. This tests the code, but the test is somewhat contrived for checking out by date since `cvs status` doesn't output anything related to dates. Of course, all dates also change every time the test is run.

I checked the remaining uncovered lines. These are functions in `CvsFetchStrategy` which are indeed not covered by tests:
- Connecting to a remote CVS server (as opposed to accessing a local repository)
- The functions `source_id`, `mirror_id`, `archive`, and `__str__`. I implemented these by following what is implemented for SVN repos. Since I don't actually understand what these do, I could remove them (and call `__str__` from a test).

adamjstewart reviewed (dismissed):
Still want to see other core devs review this PR.

eschnett commented:
ping

Is there something I should do?

eschnett commented:
Hello?

scheibelp commented:
I can look at this on Monday (June 7th)

scheibelp commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1173,6 +1174,60 @@ you can check out a branch or tag by changing the URL. If you want to
+  This download method is untrusted.
+
+  Unfortunately, CVS has no commit hashing scheme like Git and
+  Mercurial do, so there is no way to guarantee that the download you
```
Technically we don't actually trust those hashes for security either (e.g. git explicitly claims that its hashes are not for security): we support an `https:`-based clone for `svn`/`hg`/`git` which can provide a secure retrieval. So this wording should be updated.

From reading https://www.gnu.org/software/trans-coord/manual/cvs/cvs.html#Remote-repositories section 2.9.2, it looks like ssh is supported, so theoretically this could offer similar protection. Without something like that though it appears to use its own (insecure) protocol - am I correct about that?

eschnett replied:
Yes, you are correct. Most public CVS repositories support only the `pserver` method, which is the CVS-specific insecure protocol. The code thus supports only two access methods, either a local directory or `pserver`.

My intent was to say that it is not possible to specify a particular commit that should be checked out. I wasn't worried about ensuring that the commit hasn't been tampered with. I updated the wording.

scheibelp commented on lib/spack/spack/test/conftest.py:
```diff
@@ -884,6 +886,126 @@ def mock_archive(request, tmpdir_factory):
+        sourcedir.ensure(r0_file)
+        cvs('-d', cvsroot, 'add', r0_file)
+        cvs('-d', cvsroot, 'commit', '-m', 'revision 0', r0_file)
+        revision_date['1.1'] = format_date(datetime.now())
```
Can you use fixed and made-up dates with time deltas (i.e. make up some date and then add 1 hour to it)?

eschnett replied:
For this, I need to call `cvs` with that made-up time. I assume `cvs` queries the kernel to find the current date and time. Is this possible in Spack's mocking infrastructure?

scheibelp replied:
Now that I read through this more, are these generated dates actually used? It looks like they are a fallback when `get_date` should be pulling a date from `cvs`? If the act of creating `cvs` commits records timestamps, then I see why we would be working with real times, but I don't see the reason for recording these other times outside of `cvs`.

We cannot mock anything that `cvs` calls (we are only able to update the behavior of python functions). Actually strictly speaking that's not true since if it calls an exe to get this, we could override `PATH`, but I'd prefer to avoid that.

eschnett replied:
This is the code that creates a test repo. I record the times here so that I can use these same dates later in the test case to check out the repo at this date (to check checking out by date).

scheibelp replied:
Does this depend on the timestamps in the repo matching the recorded times? I think it's possible they could differ (even though they run right after one another). Could you retrieve the date from the repo as you add revisions?

eschnett replied:
No, the time stamps do not need to match. Yes, I expect them to differ from time to time, in particular on a slow file system.

When you specify a time, then cvs will look at the state of the repo at that time, i.e. it will (for each file separately) use the latest commit _not later_ than the time you specify. Thus choosing a time that is later than the latest commit will work fine.

In a realistic use case one would specify "the repo state on June 8th", not the exact date or time of the latest commit.

scheibelp replied:
OK I think I understand now. Can you add a comment explaining this e.g. to the declaration of `revision_date`:

```
# We use this to record the times when we create CVS revisions, so that we can later
# check that we retrieve the proper revision when specifying a date (CVS guarantees
# checking out the lastest revision before or on the specified date). As we create each
# revision, we separately record the time (we expect that for a sequence of CVS
# invocations and datetime.now() calls, that the resulting timestamps have the same
# ordering)
revision_date = {}
```

Unfortunately several things might screw up that last assumption. However, if there is no way to ask CVS "what is the date of the latest revision you have" then using `datetime.now` seems like an appropriate fallback (still, I want a warning there if the CI environment changes in a way that violates this assumption).

scheibelp replied:
> if there is no way to ask CVS "what is the date of the latest revision you have" then using datetime.now seems like an appropriate fallback 

I see at https://stackoverflow.com/questions/1044651/how-to-programatically-get-the-latest-commit-date-on-a-cvs-checkout there appears to be a means to retrieve the latest timestamp for a specific file. If there are problems with this test in the future related to timestamp ordering, then I think that link would be useful (although I don't think this needs to be done now).

eschnett replied:
I added a respective note in a comment. Thanks.

scheibelp commented on lib/spack/spack/test/conftest.py:
```diff
@@ -884,6 +886,126 @@ def mock_archive(request, tmpdir_factory):
+                try:
+                    rdate = revision_date[rev]
+                    date = max(date, rdate)
+                except Exception:
```
Catch something more specific (I'm assuming that it is expected that this will be hit in some cases - if you don't anticipate this exception being raised, then don't catch it at all)

eschnett replied:
This checks whether the key is in the `revision_date` dictionary. I changed it to use an if statement `key in revision_date` instead.

scheibelp commented on lib/spack/spack/test/conftest.py:
```diff
@@ -884,6 +886,126 @@ def mock_archive(request, tmpdir_factory):
+                    # First branch name found
+                    branch = tag
+                elif tag == branch:
+                    # Later branch name found; all branch names agree
```
I recommend the following wording: "all collected branch names agree so far"

eschnett replied:
Done.

scheibelp commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -940,6 +940,125 @@ def __str__(self):
+        return self.cache_enabled and (bool(self.branch) or bool(self.date))
+
+    def source_id(self):
+        id = 'id'
```
(major)

`source_id` should only return `x` such that `bool(x) == True` if the version is something that pins down a well-defined state of the repository (i.e. something that could be retrieved later after the repository has changed). It looks like if a date or branch is not specified, you are just pulling the latest state; Spack doesn't keep track of what this was, so this should return `None` if `self.branch` and `self.date` are not defined.

eschnett replied:
Thanks for the pointer. This is what reviews are for!

scheibelp commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -940,6 +940,125 @@ def __str__(self):
+                cmds = ['-c', 'spawn cvs -q -d ' + url + ' login',
+                        '-c', 'expect',
+                        '-c', 'send "anonymous\n"',
+                        '-c', 'expect eof']
```
(major)

This usage of `expect` is unexpected - according to section 2.9.3.2 of https://www.gnu.org/software/trans-coord/manual/cvs/cvs.html#Remote-repositories you should not have to log in with a password-less user.

I'd like to avoid usage of `expect` if at all possible: for example I'm not clear how these operations affect the CVS client state for subsequent commands (should we be logging out?).

eschnett replied:
There is a file `$HOME/.cvspass` that contains all the ("encrypted") passwords after a login. `cvs logout` would remove the respective password. So yes, this would change global state.

If I recall correctly, in many cases (when using `pserver`) one still has to log in before accessing a remote repository. It's just that the username/password combination is publicly posted on the download web site. See e.g. <https://lorene.obspm.fr/download.html>. At least for this repository, logging in is definitively necessary.

I see in the documentation that one can instead put the password into the download url. I'll try this instead.

scheibelp commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -940,6 +940,125 @@ def __str__(self):
+            id += '-date=' + self.date
+        return id
+
+    def mirror_id(self):
```
Like `source_id` this should only ever return non-`None` when there is a branch/date defined.

eschnett replied:
Done.

scheibelp reviewed (changes requested):
I have a few minor suggestions and a couple issues of greater concern (which I have marked as "major") relating to the retrieval and caching of these resources.

eschnett committed:
Implement all but one recommendations

scheibelp commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1173,6 +1174,62 @@ you can check out a branch or tag by changing the URL. If you want to
+
+     version('2021.4.22', branch='branchname', date='2021-04-22')
+
+  This download method is untrusted.
```
Currently you have this warning in each version section. I think it would make more sense to place it in an initial paragraph (up around line 1188) that highlights that this is an issue for all uses of `:pserver:` (or a more general message that says something like "fetching from public servers is not secured (lacks integrity guarantees of ssh/https)".

You could point users to `:extssh:` in the case that they maintain a private `cvs` if desired, but I don't think it's necessary to include that info.

eschnett replied:
I shortened the text and removed the duplication.

eschnett committed:
Simplify documentation

eschnett committed:
Correct conversion error for datetimes

eschnett committed:
Don't require Python 3.7

eschnett committed:
Add empty line

eschnett committed:
Add comment describing CVS revision -> date data structure

scheibelp commented:
Close/reopen to restart tests

scheibelp commented on lib/spack/spack/test/cvs_fetch.py:
```diff
@@ -0,0 +1,100 @@
+        source_path = stage.source_path
+        mkdirp(source_path)
+
+        fetcher.fetch()
```
I see one fetch but not two fetches (or a previous download) so I'm confused what this is testing.

eschnett replied:
I don't understand the overall fetch testing strategy in spack very well. I assumed that the `with Stage` would "stage" the repo (i.e. fetch it), and the second fetch would test it.

The other `test_*_extra_fetch` (svn, git, hg) all use an equivalent function. I was hoping that copying their approach would be the right thing to do.

scheibelp replied:
Well apologies as you've inherited some technical debt here: it doesn't look like this works to me (e.g. if I insert a debugger statement in `GitFetchStrategy.fetch` and then run `spack unit-test -k git_extra_fetch`, it doesn't hit the `GitFetchStrategy.fetch` function until the explicit call to `fetch` in the unit test). I'd say you could add a TODO here for now and we'll have to fix those other tests at some point (although perhaps just calling `.fetch` twice would test what we want).

eschnett replied:
Done.

scheibelp commented on lib/spack/spack/test/cvs_fetch.py:
```diff
@@ -0,0 +1,100 @@
+
+            # Check date
+            if test.date is not None:
+                assert get_date() <= test.date
```
I've been reading through the fixture and this part seems a bit "too mocked": we generate the timestamps ourselves in the test, then generate a later timestamp in the test, and then compare the two. Would it be possible to collect timestamps of the files from the cvs repository and check that they match what we expect?

eschnett replied:
I'm now querying CVS for the time stamp ("date"), and use this date to check out a particular set of commits.

eschnett committed:
Add note regarding improving checking timestamp ordering

eschnett committed:
Add TODO comment in test_cvs_extra_fetch

Also call `fetcher.fetch` twice for good measure.

scheibelp commented:
Thanks for adding an explanatory comment for https://github.com/spack/spack/pull/23212#discussion_r649324718

To make sure I'm in sync, there isn't yet a fix for https://github.com/spack/spack/pull/23212#discussion_r649408867 right? I was thinking the test logic should change for that.

eschnett committed:
Query CVS for file timestamps instead of assuming the current time

eschnett committed:
Use a non-trivial CVS time stamp to check checking out by time

scheibelp commented on lib/spack/docs/getting_started.rst:
```diff
@@ -23,6 +23,7 @@ before Spack is run:
    executables for extracting source code
 #. The ``patch`` command to apply patches
 #. The ``git`` and ``curl`` commands for fetching
+#. If using CVS, the ``cvs`` command for fetching
```
We don't specifically call out `go` or `hg`, so I don't think we need to call out `cvs` here either.

Other than this, the PR looks good to me (and I will approve it with that change).

eschnett replied:
True. Done.

eschnett commented:
> Thanks for adding an explanatory comment for [#23212 (comment)](https://github.com/spack/spack/pull/23212#discussion_r649324718)
> 
> To make sure I'm in sync, there isn't yet a fix for [#23212 (comment)](https://github.com/spack/spack/pull/23212#discussion_r649408867) right? I was thinking the test logic should change for that.

No, I don't think there is a fix yet. I added a second fetch as you suggested, but I don't know whether this second fetch is a no-op: There would need to be some code that takes a snapshot of the directory after the first fetch and compares it to the directory after the second fetch, and I don't see such code. If, for example, a `cvs checkout` were to leave an unwanted file with a random name around, I don't think this would be detected.

eschnett committed:
Don't document CVS-specific software requirements so prominently

eschnett committed:
Remove debug statement

eschnett committed:
Remove debug statement

eschnett committed:
Be flexible when parsing dates

eschnett committed:
Add "dateutil" package

eschnett committed:
CI: Install missing packages

eschnett committed:
Correct yaml syntax

eschnett committed:
CI: Correct pip command

eschnett committed:
Merge branch 'develop' into eschnett/cvs

# Conflicts:
#	.github/workflows/unit_tests.yaml

eschnett committed:
Remove unused package

eschnett committed:
Correct package name

eschnett committed:
More work on tests

eschnett committed:
Corrections to tests

eschnett committed:
Remove debug output

eschnett commented:
@scheibelp There has been a development after your review...

It turns out that CVS uses different date formats on different systems. Parsing with `strptime` doesn't work, and I'm using the `python-dateutils` package instead. This works fine, but this package is not available for Python 2.6. That leaves three options:
- Develop my own general date parsing function (I don't want to do this)
- Find a similar package that supports Python 2.6 (I couldn't find one)
- Disable testing "checkout by date" on Python 2.6

Note that this only affects testing, CVS support itself doesn't require this package.

I've implemented the third with a few `if` statements. See the beginning of `conftest.py`:
```Python
if sys.version_info >= (2, 7):
    # This package does not exist for Python <2.7. That means that we cannot
    # test checkouts "by date" for CVS respositories. (We can still use CVS
    # repos with all features, only our tests break.)
    from dateutil.parser import parse as parse_date
else:
    def parse_date(string):
        return None
```
Please let me know whether this is fine.

eschnett commented:
I've converted this to a draft since I didn't want this to be merged automatically.

scheibelp commented:
I also prefer option 3 of https://github.com/spack/spack/pull/23212#issuecomment-862533698, although my preference is that this be implemented with `pytest.mark.skipif` on the test itself (if that's hard to do as a decorator because of parameterization, I think `git_fetch` has an example of `pytest.skip` inside of the unit test.

On the other hand, that separates the skip from the reason (i.e. the construction in the fixture in `conftest`, so I'm wondering if you could conditionally define a `parse_date` that looks like:

```
def parse_date(string):
    pytest.skip("Not available for Python 2.6")
```

And it also occurs to me it might be worth mentioning that CVS outputs different date formats for different systems in a comment related to this logic.

eschnett committed:
Use pytest.skip for skipping tests

eschnett committed:
Add detail to comment explaining dateutil

scheibelp commented:
Closing/reopening to restart pipeline

scheibelp commented:
Thanks!

eschnett commented:
🎆🎆🎆

On Tue, Jun 22, 2021 at 12:51 PM Peter Scheibel ***@***.***>
wrote:

> Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/23212#issuecomment-866158577>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUECUDYKDGTO6Y6V7PXTDTUC5SRANCNFSM43NRZIRA>
> .
>


-- 
Erik Schnetter ***@***.***>
http://www.perimeterinstitute.ca/personal/eschnetter/

scheibelp commented on lib/spack/spack/test/conftest.py:
```diff
@@ -12,9 +12,21 @@
 import tempfile
 import xml.etree.ElementTree
 
+if sys.version_info >= (2, 7):
```
@eschnett according to https://github.com/spack/spack/pull/24471/checks?check_run_id=2887741511 this is also not supported for Python 2.7 either (at least on our CI systems). I'm not sure why but that didn't get caught in the CI for this PR, but I think this statement will have to skip import of `dateutil` for all Python 2.x. Do you have a minute to update this? I'm trying to get time ASAP but won't be able to for another hour or so.

scheibelp replied:
I've made an attempt at this at https://github.com/spack/spack/pull/24473

eschnett replied:
Thanks!

eschnett replied:
I still see errors like the one above (e.g. at https://github.com/spack/spack/pull/24388/checks?check_run_id=2890513500). I don't think this is related to Python 2.7, because it's not the CVS-related code that is failing.

I think what is going on is the following: As part of the test suite, there is an analysis stage (related to `external/_pytest`) which analyzes the Python code. This analysis stage comes across the `import dateutil` statement and cannot handle it.

I don't know what `external/_pytest` does, or what analysis stage that is. But maybe it would suffice to tell this stage that `dateutil` is now a fine package to be imported, and to not worry about it.

scheibelp replied:
I am surprised: at least one PR submitted since #24473 has passed CI (https://github.com/spack/spack/pull/24481) so I'm wondering if there is an issue where PRs that failed due to this are difficult to correct (for the PR you linked, it hit this problem, and you pushed additional commits after it was corrected, but I see the CI fails in both cases).

alalazo replied:
Posting a detail that I noticed on a PR of mine #24398 that is being hit by the same issue. When the error happens it seems Spack picks Python 3.8 as the interpreter, even if it is in the Python 2.7 CI job. See [here](https://github.com/spack/spack/pull/24398/checks?check_run_id=2892195010) for the log of the failure.

alalazo replied:
I think this uncovered some kind of subtle bug where package only PRs are running on Python 3.8 when in the Python 2.7 CI job

alalazo commented:
I see this PR introduced a dependency on `python-dateutil` at least to run tests, but if I am not missing something it doesn't vendor the corresponding package. This means that before this PR one could git clone Spack and run:
```console
$ spack unit-test
```
while after this PR we need to:
```console
$ pip install python-dateutil
```
before running any unit test locally. Is this wanted?

eschnett commented:
@alalazo I don't know what "to vendor" means.

Yes, one now needs to install `python-dateutils` as well – I didn't realize that this could lead to trouble. I guess it would be possible to test whether this package is installed, and if not, skip the respective tests with a warning?

alalazo commented:
> I don't know what "to vendor" means.

That's a [good definition](https://stackoverflow.com/questions/26217488/what-is-vendoring) of what I mean by "vendor". We have a few 3rd party libraries shipped with Spack in `lib/spack/spack/external`.

> I guess it would be possible to test whether this package is installed, and if not, skip the respective tests with a warning?

Should be hopefully solved by https://github.com/spack/spack/pull/24484 thanks!
